### PR TITLE
remove SharedDDP as it is deprecated

### PR DESCRIPTION
**Pull request by ji-huazhong**
# What does this PR do?

As mentioned previously([see](https://github.com/huggingface/transformers/pull/24825)), fairscale's ShardedDDP is deprecated,  and PyTorch FSDP is the recommended method for scaling to large NN models. Now it's time to say goodbye to this library👋. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@muellerz Good day. Could you please review this PR? Thanks😄

### CODE REVIEW

File: src/transformers/trainer.py
Comment (statelesshz) [inline src/transformers/trainer.py:627]: mix precision logic is all moved to accelerate, so these codes do not take effect any longer when removing SharededDDP.

File: src/transformers/trainer.py
Comment (statelesshz) [inline src/transformers/trainer.py:1324]: I'm curious as to whether the codes are still necessary now that Accelerate supports model wrapping with IPEX.

https://github.com/huggingface/accelerate/blob/81b19c4094395db1be11150a900a92504190f7ac/src/accelerate/accelerator.py#L1260-L1264

File: src/transformers/trainer.py
Comment (statelesshz) [inline src/transformers/trainer.py:2642]: as above, autocast_smart_context_manager does not take effect any longer when removing SharededDDP.

**Comment by statelesshz** [comment]
It seems that the failing test case is not related to this PR.

**Comment by statelesshz** [comment]
@sgugger sorry for bother, would you mind take a look at this PR?

**Comment by sgugger** [comment]
cc @muellerzr and @pacman100 who will take over the Trainer.

**Comment by statelesshz** [comment]
@muellerzr and @pacman100 please take a look at this PR, thx :-)

**Comment by statelesshz** [comment]
Rebase my commits to master HEAD to fix the merge conflict.
BTW, Is this PR still under reviewing? Any review suggestions? Please let me know if there is anything else that needs to be done.

File: src/transformers/trainer.py
Comment (pacman100) [reply src/transformers/trainer.py:2642]: if `autocast_smart_context_manager ` isn't being used, we can remove the function altogether?

File: src/transformers/trainer.py
Comment (pacman100) [inline src/transformers/trainer.py:1878]: cc @muellerzr as you have more idea about TPUs

File: src/transformers/trainer.py
Comment (pacman100) [inline src/transformers/trainer.py:1908]: cc @muellerzr as you have more idea about TPUs

**Review by pacman100 — COMMENTED** [review]
Thank you @statelesshz for removing the FairScale support as it is deprecated. The changes LGTM except for the TPU-related bits which Zach can review. Left a comment.

File: src/transformers/trainer.py
Comment (statelesshz) [reply src/transformers/trainer.py:2642]: fixed in [d2e01a](https://github.com/huggingface/transformers/pull/25702/commits/d2e01a5922f53eadf147d551c240dbf226a6cc60)

Comment (statelesshz) [reply src/transformers/trainer.py:2642]: fixed

File: src/transformers/trainer.py
Comment (statelesshz) [reply src/transformers/trainer.py:1878]: @muellerzr Could you please take a look at these modifications？Thanks :-)

Comment (muellerzr) [reply src/transformers/trainer.py:1880]: This XLA logic needs to be maintained in here, or have the equivalent done in Accelerate in terms of the `all_reduce` + `fetch_gradients`. If you'd like to open a PR for this that'd be phenomenal! (Sorry, had to go dig in accelerate for a bit to find the relevant areas)

File: src/transformers/trainer.py
Comment (muellerzr) [inline src/transformers/trainer.py:1785]: Correct me if I'm wrong, but we can now remove this `if` entirely and just have `self.optimizer.step` and the check for if it were skipped?

**Review by muellerzr — COMMENTED** [review]
Thanks! This is looking great! Just a small note in terms of gradient scaling. We need to either keep this for now in this PR and merge, or coordinate a PR in accelerate with the right logic before this merges. I don't particularly have a leaning one way or another :)

**Comment by statelesshz** [comment]
rebase my commits from the main branch

File: src/transformers/trainer.py
Comment (statelesshz) [reply src/transformers/trainer.py]: Nice catch!

File: src/transformers/trainer.py
Comment (statelesshz) [reply src/transformers/trainer.py:1880]: Once I figure out how it works, I would be happy to have the equivalent done in Accelerate, but it might take some time :-).

**Comment by statelesshz** [comment]
@muellerzr Could you please take a second look at this PR? 
Some modifications were made based on the code review comments.

**Review by muellerzr — APPROVED** [review]
Thanks! This all looks great to me, do you have access to a multi-gpu system by chance to run the sharded DDP yourself/test? Probably also good to have @pacman100 give it one more look as well :)

**Comment by statelesshz** [comment]
I've tested this PR using 4xA100-80G on [FastChat](https://github.com/lm-sys/FastChat) with the following scripts
```
CUDA_VISIBLE_DEVICES=4,5,6,7  torchrun --nproc_per_node=4 --master_port=20001 fastchat/train/train_mem.py \
    --model_name_or_path lmsys/vicuna-7b-v1.5  \
    --data_path data/dummy_conversation.json \
    --bf16 True \
    --output_dir output_vicuna \
    --num_train_epochs 1 \
    --max_steps 10 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 1 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 1200 \
    --save_total_limit 10 \
    --learning_rate 2e-5 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --fsdp "full_shard offload auto_wrap" \
    --fsdp_transformer_layer_cls_to_wrap 'LlamaDecoderLayer' \
    --tf32 True \
    --model_max_length 2048 \
    --gradient_checkpointing False \
    --lazy_preprocess True

```

**Comment by statelesshz** [comment]
@pacman100 Could you please take a look once again :-) I think it's ready to be merged.

File: src/transformers/trainer.py
Comment (statelesshz) [reply src/transformers/trainer.py:1880]: remove XLA logic, since Accelerate will handle it automatically.

**Comment by statelesshz** [comment]
Resolve merge conflicts by rebasing to master branch

File: src/transformers/trainer.py
Comment (pacman100) [reply src/transformers/trainer.py:1325]: Hello, l think IPEX is still not supported via Accelerate. Let's not touch IPEX parts in this PR.

**Review by pacman100 — COMMENTED** [review]
Thank you for iterating! Left a comment

File: src/transformers/trainer.py
Comment (pacman100) [reply src/transformers/trainer.py:1325]: I think this comment still needs to be resolved

**Comment by statelesshz** [comment]
@pacman100 Thanks for your comments and patience, I've reverted the modifications to the IPEX logic

File: src/transformers/trainer.py
Comment (statelesshz) [reply src/transformers/trainer.py]: Oh, thanks for pointing this out. fixed

File: src/transformers/trainer.py
Comment (muellerzr) [reply src/transformers/trainer.py:1879]: This has been merged into main on Accelerate via https://github.com/huggingface/accelerate/pull/1926

**Review by muellerzr — APPROVED** [review]
This looks good to me then, users should have been getting the future warnings since July which ought to be enough of a heads up with this change. WDYT @pacman100 @amyeroberts

File: src/transformers/trainer.py
Comment (amyeroberts) [inline src/transformers/trainer.py:1797]: Is this guaranteed to be the same if `is_torch_tpu_available()` is `True`? Previously, `optimizer_was_run` was always `True`

File: src/transformers/trainer.py
Comment (amyeroberts) [inline src/transformers/trainer.py:572]: Is this change OK to do? Previously we were only setting `args.half_precision_backend` if `self.sharded_ddp` was set (I'm guessing yes because of the removal of "cuda_amp" as an option)

@muellerzr You'll have a better idea of how this is managed :)

**Review by amyeroberts — COMMENTED** [review]
Thanks for tidying up the code and handling this deprecation! 

I just have two small questions. Otherwise changes look good - great to see so much code removed!

File: src/transformers/trainer.py
Comment (muellerzr) [reply src/transformers/trainer.py:1797]: Yes, this will always still be true. In accelerate it's only false on non-tpu when grad scaler comes into play.

File: src/transformers/trainer.py
Comment (statelesshz) [reply src/transformers/trainer.py:572]: I prefer keeping the current logic for now and opening another PR to handle logic related to the training argument `half_precision_backends` :-)  WDYT @muellerzr

Comment (amyeroberts) [reply src/transformers/trainer.py:572]: @statelesshz We can't merge in a PR that will possibly silently change behaviour and then bring in another PR later which rectifies. If this is done in two steps, then the second PR you propose would need to be added first.

**Comment by statelesshz** [comment]
After further investigation, I think the current modification is acceptable.
FYI, https://github.com/huggingface/transformers/pull/23148 shifts `Trainer`'s mixed precision handling to accelerate. From https://github.com/huggingface/transformers/pull/23148#issue-1695779875 we know, `(fairscale)Sharded DDP and Apex are cases not supported in Accelerate `, so @pacman100 inserted `self.sharded_ddp is not None` at line 617 of `trainer.py`

And in this PR, we will remove fairscale ShardedDDP completely and the mixed precision logic needs to be handled synchronously.
Firstly I removed the `use_cuda_amp` related logic since Accelerate will handle it automatically. Secondly, we keep the logic of the Apex as Apex is not supported by Accelerate. Finally, I was going to remove use_cpu_amp but ended up keeping it

File: src/transformers/trainer.py
Comment (statelesshz) [reply src/transformers/trainer.py]: After further investigation, I think the current modification for mixed precision setup is acceptable.
FYI, https://github.com/huggingface/transformers/pull/23148 shifts `Trainer`'s mixed precision handling to accelerate. From https://github.com/huggingface/transformers/pull/23148#issue-1695779875 we know, `(fairscale)Sharded DDP and Apex are cases not supported in Accelerate `, so @pacman100 inserted `self.sharded_ddp is not None` at line 617 of `trainer.py`

And in this PR, we will remove fairscale ShardedDDP completely and the mixed precision logic needs to be handled synchronously.
Firstly I removed the `use_cuda_amp` related logic since Accelerate will handle it automatically. Secondly, we keep the logic of the Apex as Apex is not supported by Accelerate. Finally, I was going to remove use_cpu_amp but ended up keeping it

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing and iterating on this! 

Once tests are passing, I think we're good to merge

**Review by pacman100 — APPROVED** [review]
Thank you @statelesshz for iterating, LGTM for merging once the tests are green

**Comment by statelesshz** [comment]
Rebasing my commits to master HEAD and resolving merge conflicts

**Comment by statelesshz** [comment]
@muellerzr @pacman100 @amyeroberts Good day. I think this PR is ready to be merged：）

**Comment by statelesshz** [comment]
Hi there. This PR is approved and the tests are green :D
cc @muellerzr and @pacman100

**Comment by statelesshz** [comment]
This PR is approved and the tests are green. @muellerzr Could you help to merge it?

**Comment by LysandreJik** [comment]
Thank you @statelesshz!

**Comment by LysandreJik** [comment]
Let me rebase quickly and merge if tests are green

**Review by LysandreJik — APPROVED** [review]
Great, LGTM! Thanks @statelesshz

**Pull request merged**
